File: labs/lab5/PyBelkin/Messages.py
from dataclasses import dataclass
import socket, struct


@dataclass
class MessageTypesBelkin:
    MT_CONNECT: int = 0
    MT_DISCONNECT: int = 1
    MT_INFO: int = 2
    MT_DATA: int = 3


@dataclass
class MessageHeaderBelkin:
    Type: int = MessageTypesBelkin.MT_INFO
    Size: int = 0
    To: int = -2
    From: int = 0


class MessageBelkin:
    def __init__(self, To=-2, From=0, Type=MessageTypesBelkin.MT_INFO, Data=""):
        self.Header = MessageHeaderBelkin(Type, len(Data) * 2, To, From)
        self.Data = Data

    def __str__(self):
        return f"[{self.Header.From}]: {self.Data}"

    def send(self, s: socket.socket):
        s.send(struct.pack(f"iiii", self.Header.Type, self.Header.Size, self.Header.To, self.Header.From))

        if self.Header.Size > 0:
            s.send(struct.pack(f'{self.Header.Size}s', self.Data.encode('utf-16-le')))

    def receive(self, s: socket.socket) -> bool:
        try:
            (self.Header.Type, self.Header.Size, self.Header.To, self.Header.From) = struct.unpack('iiii', s.recv(16))

            if self.Header.Size > 0:
                self.Data = struct.unpack(f'{self.Header.Size}s', s.recv(self.Header.Size))[0].decode('utf-16-le')

            return True
        except Exception:
            return False

File: labs/lab5/SharpBelkin/Messages.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Net.Sockets;
using System.Runtime.InteropServices;
using System.Text;

namespace SharpBelkin
{
    public enum MessageTypesBelkin : int
    {
        MT_CONNECT = 0,
        MT_DISCONNECT = 1,
        MT_INFO = 2,
        MT_DATA = 3
    }

    [StructLayout(LayoutKind.Sequential)]
    struct MessageHeaderBelkin
    {
        [MarshalAs(UnmanagedType.I4)]
        public MessageTypesBelkin type;
        [MarshalAs(UnmanagedType.I4)]
        public int size;
        [MarshalAs(UnmanagedType.I4)]
        public int to;
        [MarshalAs(UnmanagedType.I4)]
        public int from;
    }


    class MessageBelkin
    {
        public MessageHeaderBelkin header;
        public string data;

        public MessageBelkin(int to = -2, int from = 0, MessageTypesBelkin type = MessageTypesBelkin.MT_INFO, string data = "")
        {
            this.data = data;
            header = new MessageHeaderBelkin() { type = type, size = data.Length * 2, to = to, from = from };
        }

        static byte[] ToBytes(object obj)
        {
            int size = Marshal.SizeOf(obj);
            byte[] buff = new byte[size];
            IntPtr ptr = Marshal.AllocHGlobal(size);
            Marshal.StructureToPtr(obj, ptr, true);
            Marshal.Copy(ptr, buff, 0, size);
            Marshal.FreeHGlobal(ptr);
            return buff;
        }

        static T fromBytes<T>(byte[] buff) where T : struct
        {
            T data = default(T);
            int size = Marshal.SizeOf(data);
            IntPtr i = Marshal.AllocHGlobal(size);
            Marshal.Copy(buff, 0, i, size);
            var d = Marshal.PtrToStructure(i, data.GetType());
            if (d is not null)
            {
                data = (T)d;
            }
            Marshal.FreeHGlobal(i);
            return data;
        }

        public void send(Socket s)
        {
            s.Send(ToBytes(header), Marshal.SizeOf(header), SocketFlags.None);
            if (header.size != 0)
            {
                s.Send(Encoding.Unicode.GetBytes(data), header.size, SocketFlags.None);
            }
        }

        public bool receive(Socket s)
        {
            try 
            {
                byte[] buff = new byte[Marshal.SizeOf(header)];
                if (s.Receive(buff, Marshal.SizeOf(header), SocketFlags.None) == 0)
                {
                    return false;
                }
                header = fromBytes<MessageHeaderBelkin>(buff);
                if (header.size > 0)
                {
                    byte[] b = new byte[header.size];
                    s.Receive(b, header.size, SocketFlags.None);
                    data = Encoding.Unicode.GetString(b, 0, header.size);

                }
                return true;
            }
            catch
            {
                return false;
            }
            
        }
    }
}
